modules/core/shared/src/main/scala-3/io/circe/derivation/package.scala

```diff
@@ -25,28 +25,70 @@ private[circe] inline final def summonLabels[T <: Tuple]: List[String] =
+private[circe] inline final def summonEncoders[T <: Tuple](inline autoRecurse: Boolean)(using
+  Configuration
+): List[Encoder[_]] =
   inline erasedValue[T] match
     case _: EmptyTuple => Nil
-    case _: (t *: ts)  => summonEncoder[t] :: summonEncoders[ts]
+    case _: (t *: ts)  => summonEncoder[t](autoRecurse) :: summonEncoders[ts](autoRecurse)
 
-private[circe] inline final def summonEncoder[A](using Configuration): Encoder[A] =
+@deprecated("Use summonEncoders(autoRecurse: Boolean) instead", "0.14.6")
+private[circe] inline final def summonEncoders[T <: Tuple](using Configuration): List[Encoder[_]] = summonEncoders(true)
+
+private[circe] inline final def summonEncoderAutoRecurse[A](using Configuration): Encoder[A] =
   summonFrom {
     case encodeA: Encoder[A] => encodeA
     case _: Mirror.Of[A]     => ConfiguredEncoder.derived[A]
   }
 
-private[circe] inline final def summonDecoders[T <: Tuple](using Configuration): List[Decoder[_]] =
+private[circe] inline final def summonEncoderNoAutoRecurse[A](using conf: Configuration): Encoder[A] =
+  summonFrom {
+    case encodeA: Encoder[A] => encodeA
+    case m: Mirror.Of[A] =>
+      inline m match {
+        case _: Mirror.Singleton => ConfiguredEncoder.derivedNoAutoRecursion[A]
+        case _: Mirror           => error("Failed to find an instance of Encoder[" + constValue[m.MirroredLabel] + "]")
+      }
+  }
+
+private[circe] inline final def summonEncoder[A](inline autoRecurse: Boolean)(using Configuration): Encoder[A] =
+  inline if (autoRecurse) summonEncoderAutoRecurse[A] else summonEncoderNoAutoRecurse[A]
+
+@deprecated("Use summonEncoder(autoRecurse: Boolean) instead", "0.14.6")
+private[circe] inline final def summonEncoder[A](using Configuration): Encoder[A] = summonEncoder(true)
+
+private[circe] inline final def summonDecoders[T <: Tuple](inline autoRecurse: Boolean)(using
+  Configuration
+): List[Decoder[_]] =
   inline erasedValue[T] match
     case _: EmptyTuple => Nil
-    case _: (t *: ts)  => summonDecoder[t] :: summonDecoders[ts]
+    case _: (t *: ts)  => summonDecoder[t](autoRecurse) :: summonDecoders[ts](autoRecurse)
 
-private[circe] inline final def summonDecoder[A](using Configuration): Decoder[A] =
+@deprecated("Use summonDecoders(autoRecurse: Boolean) instead", "0.14.6")
+private[circe] inline final def summonDecoders[T <: Tuple](using Configuration): List[Decoder[_]] = summonDecoders(true)
+
+private[circe] inline final def summonDecoderAutoRecurse[A](using Configuration): Decoder[A] =
   summonFrom {
     case decodeA: Decoder[A] => decodeA
     case _: Mirror.Of[A]     => ConfiguredDecoder.derived[A]
   }
 
+private[circe] inline final def summonDecoderNoAutoRecurse[A](using conf: Configuration): Decoder[A] =
+  summonFrom {
+    case decodeA: Decoder[A] => decodeA
+    case m: Mirror.Of[A] =>
+      inline m match {
+        case _: Mirror.Singleton => ConfiguredDecoder.derivedNoAutoRecursion[A]
```

Comment:
Isn't Mirror.Singleton only available for case objects?

Reply:
Yeah, this is intentional. From the PR description:

> the *NoAutoRecursion variants will still auto-recurse for types with a Mirror.Singleton instance

and from the description in #2173 

> auto-recursion will only happen for types with a `Mirror.Singleton` instance, i.e. `case object`s. This allows for code like this to still compile:

```scala
sealed trait Foo
case object Bar extends Foo

given conf: Configuration = Configuration.default.withoutAutoRecursiveDerivation

// Compiles even though we haven't defined an `Encoder[Bar.type]`
Encoder.derivedConfigured[Foo]
```

Reply:
Adding some tests showing what does and does not auto-recurse (including compiles and does not compile tests) would help make this more clear, and prevent future regressions.

Reply:
Good idea @zarthross, I just added a test suite to demonstrate what compiles and what doesn't